bip-0322.mediawiki

```diff
@@ -13,119 +13,99 @@
+    nVersion = 0 or as appropriate (e.g. 2, for time locks)
+    nLockTime = 0 or as appropriate (for time locks)
+    vin[0].prevout.hash = to_spend.txid
+    vin[0].prevout.n = 0
+    vin[0].nSequence = 0 or as appropriate (for time locks)
+    vin[0].scriptWitness = message_signature
+    vout[0].nValue = 0
+    vout[0].scriptPubKey = OP_RETURN
 
-For both cases, generate a sighash based on the given scriptPubKey and message as follows:
+When a proof of funds is being created, additional inputs should be included for virtually spending transaction outputs of desired value.
 
-# Define the message pre-image as the sequence "Bitcoin Signed Message:\n" concatenated with the message, encoded in UTF-8 using Normalization Form Compatibility Decomposition (NFKD)
-# Let sighash = sha256(sha256(scriptPubKey || pre-image))
+* All signatures must use the SIGHASH_ALL flag.
+* The proof is considered valid, inconclusive, or invalid based on whether the to_sign transaction is a valid spend of the to_spend transaction or not, according to the rules specified in the "Consensus and standard flags" section below.
+* It may (optionally) be encumbered with the in_future flag, according to the rules specified in the "Locktime and Sequence" section below, in which case we refer to the result in text form as "valid_in_future", "inconclusive_in_future", etc.
 
-A private key may be used directly to sign a message. In this case, its P2WPKH bech32 address shall be derived, and used as the input.
+Proofs with additional inputs are the base64-encoding of the to_spend and to_sign transactions concatenated in standard network serialisation, and proofs without additional inputs (simple proofs) are the base64-encoding of the to_sign script witness.
 
-=== Signing ===
+A validator must verify it is valid and meets the description of virtual transactions as specified above. See "Validation" below.
 
-The signature is generated as follows:
+=== Validation ===
 
-# Derive the private key privkey for the scriptPubKey; FAIL if not VALID
-# Generate and return a signature sig with privkey=privkey, sighash=sighash
-
-=== Verifying ===
+To validate a proof with additional inputs, the following steps must be taken:
 
-Verify a proof, given a standard flags value, a script sig, an optional witness, and a derived sighash as described above.
+# deserialize the to_spend and to_sign transactions from the proof, and fail if the proof contains extraneous bytes
+# verify that any additional inputs being proven (proof of funds) are included as inputs in the to_sign transaction, exactly once
+# reconstruct the to_spend' and to_sign' transactions, based on the specification above, copying the version, lock time, and sequence values
+# verify that to_spend = to_spend', that to_sign has at least 1 input, has exactly 1 output, and that to_sign.vin[0] = to_sign'.vin[0]
+# optional: set the "in_future" flag if the transaction's lock time is in the future according to consensus rules
```

Comment:
I would propose that we choose to either not cover locktime at all (i.e. have the spec fail to make a determination), or explicitly state how it should be treated. I would worry that optional features in a standard would otherwise lead to variable behavior that could cause compatibility issues later on.

Reply:
I think it may make sense to not make this optional now that we've split out into two types of proofs (simple proofs and PoF).